server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/view/cassandra/CassandraMailQueueView.java

```diff
@@ -99,11 +100,20 @@ public long getSize() {
 
     @Override
     public CompletableFuture<Long> delete(DeleteCondition deleteCondition) {
-        throw new NotImplementedException("Not implemented yet");
+        CompletableFuture<Long> result = cassandraMailQueueBrowser.browseReferences(mailQueueName)
+            .filter(mailReference -> deleteCondition.shouldBeDeleted(mailReference.getMail()))
```

Comment:
do we really need to filter already deleted elements, or can we consider we are idempotent if we delete them again in the storage?

Reply:
IMO you don't understand what happens here:

We check if the browsed email matches the delete query in order to only remove matching email. 

(it is not an optimisation to avoid double delete)

Reply:
oh yes, sorry, read too fast :+1: